orm/model_bucket.go

```diff
@@ -0,0 +1,174 @@
+package orm
+
+import (
+	"reflect"
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/errors"
+)
+
+// TODO
+// - migrations
+// - do not use Bucket but directly access KVStore
+// - register for queries
+
+// Model is impelemented by any entity that can be stored using ModelBucket.
+//
+// This is the same interface as CloneableData. Using the right type names
+// provides an easier to read API.
+type Model interface {
+	weave.Persistent
+	Validate() error
+	Copy() CloneableData
+}
+
+// ModelSlicePtr represents a pointer to a slice of models. Think of it as
+// *[]Model Because of Go type system, using []Model would not work for us.
+// Instead we use a placeholder type and the validation is done during the
+// runtime.
+type ModelSlicePtr interface{}
+
+// ModelBucket is implemented by buckets that operates on Models rather than
+// Objects.
+type ModelBucket interface {
```

Comment:
This is a nice clean interface.
Missing the various Iterator/Range calls, but best to start here and add them when we update a module that uses them